helix-core/src/main/java/org/apache/helix/HelixManagerProperty.java

```diff
@@ -20,54 +20,82 @@
  */
 
 import java.util.Properties;
+
 import org.apache.helix.model.CloudConfig;
+import org.apache.helix.zookeeper.api.client.RealmAwareZkClient;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+
 /**
- * Hold Helix manager properties. The manager properties further hold Helix cloud properties
- * and some other properties specific for the manager.
+ * HelixManagerProperty is a general property/config object used for HelixManager creation.
  */
 public class HelixManagerProperty {
   private static final Logger LOG = LoggerFactory.getLogger(HelixManagerProperty.class.getName());
   private String _version;
   private long _healthReportLatency;
   private HelixCloudProperty _helixCloudProperty;
+  private RealmAwareZkClient.RealmAwareZkConnectionConfig _zkConnectionConfig;
+  private RealmAwareZkClient.RealmAwareZkClientConfig _zkClientConfig;
 
   /**
+   * ** Deprecated - HelixManagerProperty should be a general property/config object used for
+   * HelixManager creation, not tied only to Properties or CloudConfig **
+   *
    * Initialize Helix manager property with default value
    * @param helixManagerProperties helix manager related properties input as a map
    * @param cloudConfig cloudConfig read from Zookeeper
    */
+  @Deprecated
   public HelixManagerProperty(Properties helixManagerProperties, CloudConfig cloudConfig) {
     _helixCloudProperty = new HelixCloudProperty(cloudConfig);
     setVersion(helixManagerProperties.getProperty(SystemPropertyKeys.HELIX_MANAGER_VERSION));
     setHealthReportLatency(
         helixManagerProperties.getProperty(SystemPropertyKeys.PARTICIPANT_HEALTH_REPORT_LATENCY));
   }
 
+  public HelixManagerProperty() {
```

Comment:
I think the plan is only allowing the property construct through HelixPropertyFactory. Shall we just make this constructor package-private? Also the other set methods.

@zhangmeng916 Please correct me if this is not the plan.

Reply:
I don't like that idea - it loses flexibility. What is the benefit of making this package-private?

Reply:
So all the config read/processing work is in HelixPropertyFactory. In this case, we won't have unexpected overwrite or misusage of the property object. The flexibility should be in HelixPropertyFactory only.

Reply:
Then that in and of itself is a verbose way of doing things. At that point, we should eliminate and just add a HelixProperty Builder that allows creation of HelixManagerProperty instance with final fields.

Reply:
The original plan was to let users only get the manager property through `getHelixManagerProperty` with default value, and then allow them to make modification to override the fields in it. I remember there was some concern about users accidentally changing it. We can revisit the assumption though.

Reply:
Update: the factory and builder are all serving different purposes and might be useful to have in the future. We decided to keep both.